cmd/maya-apiserver/cstor-operator/storagepool_create.go

```diff
@@ -95,55 +152,80 @@ func poolCreateWorker(pool *apis.CasPool) error {
 	}
 	if len(nodeDisks.Disks.Items) == 0 {
 		return nil, errors.New("aborting storagepool create operation as no disk was found")
 	}
-	// For each of the disk, extract the device Id and fill the 'DeviceId' field of the CasPool object with it.
-	// In case, device Id is not available, fill the 'DeviceId' field of the CasPool object with device path.
-	for _, v := range nodeDisks.Disks.Items {
-		gotDisk, err := newClientSet.oecs.OpenebsV1alpha1().Disks().Get(v, metav1.GetOptions{})
-		if err != nil {
-			return nil, errors.Wrapf(err, "Failed to get device id for disk:failed to list the disks")
+
+	// Fill the node name to the CasPool object.
+	casPool.NodeName = nodeDisks.NodeName
+	//casPool.DiskList = nodeDisks.Disks.Items
+	//TODO: Improve Following Code
+	if spc.Spec.PoolSpec.PoolType == string(apis.PoolTypeStripedCPV) {
+		for _, disk := range nodeDisks.Disks.Items {
+			var diskList []apis.CspDisk
+			var group apis.DiskGroup
+			disk := apis.CspDisk{
+				Name:        disk,
+				InUseByPool: true,
+			}
+			devID, err := pc.getDeviceID(disk.Name)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to get dev is for disk %s for spc %s", disk.Name, spc.Name)
+			}
+			disk.DeviceID = devID
+			diskList = append(diskList, disk)
+			group = apis.DiskGroup{
+				Item: diskList,
+			}
+			casPool.DiskList = append(casPool.DiskList, group)
+		}
+		return casPool, nil
+	}
+	count := nodeselect.DefaultDiskCount[spc.Spec.PoolSpec.PoolType]
+	for i := 0; i <= len(nodeDisks.Disks.Items)/count; i = i + count {
+		var diskList []apis.CspDisk
+		var group apis.DiskGroup
+		for j := 0; j < count; j++ {
+
+			disk := apis.CspDisk{
+				Name:        nodeDisks.Disks.Items[i+j],
+				InUseByPool: true,
+			}
+			devID, err := pc.getDeviceID(disk.Name)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to get dev is for disk %s for spc %s", disk.Name, spc.Name)
+			}
+			disk.DeviceID = devID
+			diskList = append(diskList, disk)
 		}
-		if len(gotDisk.Spec.DevLinks) != 0 && len(gotDisk.Spec.DevLinks[0].Links) != 0 {
-			// Fill device Id of the disk to the CasPool object.
-			casPool.DeviceID = append(casPool.DeviceID, gotDisk.Spec.DevLinks[0].Links[0])
-		} else {
-			// Fill device path of the disk to the CasPool object.
-			// ToDo: Decide -- DeviceId and DevicePath fields for CasPool object.
-			// ToDO: Having these two fields for CasPool object can yield complex run tasks.
-			casPool.DeviceID = append(casPool.DeviceID, gotDisk.Spec.Path)
+		group = apis.DiskGroup{
+			Item: diskList,
 		}
+		casPool.DiskList = append(casPool.DiskList, group)
 	}
-	// Fill the node name to the CasPool object.
-	casPool.NodeName = nodeDisks.NodeName
-	// Fill the disks attached to this node to the CasPool object.
-	casPool.DiskList = nodeDisks.Disks.Items
 	return casPool, nil
 }
+
+// TODO: Move to disk package
+func (pc *PoolCreateConfig) getDeviceID(diskName string) (string, error) {
+	var deviceID string
```

Comment:
This business logic will move to `pkg/disk/v1alpha1/`
```go
// in pkg/disk/v1alpha1/kubernetes.go
func (k *Kubeclient) Get(name string, opts metav1.GetOptions{}) (*apis.Disk, error) {
  // ...
}
```
```go
// in pkg/disk/v1alpha1/disk.go
type Disk struct {
  object *apis.Disk
}

func (d *Disk) GetDeviceID() string {
  if len(d.object.Spec.DevLinks) != 0 && len(d.object.Spec.DevLinks[0].Links) != 0 {
    deviceID = d.object.Spec.DevLinks[0].Links[0]
  } else {
    deviceID = d.object.Spec.Path
  }
  return deviceID
}

type Builder struct {
  disk *Disk
  errors []error
}

func BuilderForAPIObject(d *apis.Disk) *Builder {
  // other validations 
  // like check d == nil, etc
  return &Builder{disk: &Disk{object: d}}
}

func (b *Builder) Build() (*Disk, error) {
  // other logic e.g. predicates, validations, etc
  return b.disk, nil
}

func (b *Builder) GetDeviceID() (string, error) {
  d, err := b.Build()
  if err != nil {
    return "", err
  }
  return d.GetDeviceID(), nil
}
```

```go
// In caller code
d, err := disk.KubeClient.Get(name, opts)
if err != nil {
  return err
}

id, err := disk.BuilderForAPIObject(d).GetDeviceID()
```

Reply:
Next PR.